cpp/src/strings/utilities.cu

```diff
@@ -142,14 +143,50 @@ __device__ character_flags_table_type
 __device__ character_cases_table_type character_cases_table[sizeof(g_character_cases_table)];
 __device__ special_case_mapping character_special_case_mappings[sizeof(g_special_case_mappings)];
 
-// initialization mutexes
-std::mutex g_flags_table_mutex;
-std::mutex g_cases_table_mutex;
-std::mutex g_special_case_mappings_mutex;
+// This template is a thin wrapper around per-context singleton objects.
+// It maintains a single object for each CUDA context.
+template <typename TableType>
+class per_context_cache {
+ public:
+  // Find an object cached for a current CUDA context.
+  // If there is no object available in the cache, it calls the initializer
+  // `init` to create a new one and cache it for later uses.
+  template <typename Initializer>
+  TableType* find_or_initialize(const Initializer& init)
+  {
+    CUcontext c;
+    cuCtxGetCurrent(&c);
+    auto finder = cache_.find(c);
+    if (finder == cache_.end()) {
+      TableType* result = init();
+      cache_[c]         = result;
+      return result;
+    } else
+      return finder->second;
+  }
+
+ private:
+  std::unordered_map<CUcontext, TableType*> cache_;
+};
+
+// This template is a thread-safe version of per_context_cache.
+template <typename TableType>
+class thread_safe_per_context_cache : public per_context_cache<TableType> {
```

Comment:
Is there a case when we would need a non-thread-safe version of this class? If not, we could roll them in to one.

Reply:
Like I said in [this comment](https://github.com/rapidsai/cudf/pull/5336#discussion_r433390350), a proper initialization call, which I hope would be added to the code, may hold a lock and initialize all the global states like these lookup tables while holding that lock (or it may even do without any lock if single-threaded). Then, we would want to use the thread-unsafe version both in the initialization and afterwards.